src/storage-client/src/controller.rs

```diff
@@ -880,82 +889,162 @@ where
-            };
+                };
 
-            let collection_state = CollectionState::new(
-                description.clone(),
-                since_handle.since().clone(),
-                write.upper().clone(),
-                metadata.clone(),
-            );
+                let cs = CollectionState::new(
+                    description.clone(),
+                    since_handle.since().clone(),
+                    write.upper().clone(),
+                    metadata.clone(),
+                );
+                Ok::<_, anyhow::Error>((id, description, write, since_handle, cs))
+            })
+            // Poll each future for each collection concurrently, maximum of 50 at a time.
+            .buffer_unordered(50)
+            // HERE BE DRAGONS:
+            //
+            // There are at least 2 subtleties in using `FuturesUnordered` (which
+            // `buffer_unordered` uses underneath:
+            // - One is captured here <https://github.com/rust-lang/futures-rs/issues/2387>
+            // - And the other is deadlocking if processing an OUTPUT of a `FuturesUnordered`
+            // stream attempts to obtain an async mutex that is also obtained in the futures
+            // being polled.
+            //
+            // Both of these could potentially be issues in all usages of `buffer_unordered` in
+            // this method, so we stick the standard advice: only use `try_collect` or
+            // `collect`!
+            .try_collect()
+            .await?;
 
+        let mut to_migrate = vec![];
+        // This work mutates the controller state, so must be done serially. Because there
+        // is no io-bound work, its very fast.
+        for (id, description, write, since_handle, collection_state) in to_register {
             self.state.persist_write_handles.register(id, write);
             self.state.persist_read_handles.register(id, since_handle);
 
             self.state.collections.insert(id, collection_state);
 
-            self.migrate_remap_shard(id, &description.data_source).await;
+            to_migrate.push((id, description));
+        }
 
-            self.register_shard_mapping(id).await;
+        // Reborrow `&mut self` immutably, same reasoning as above.
+        let this = &*self;
+        // We now register the shard mapping for each collection, which is effectively appending
+        // some updates to `CollectionManager` channel concurrently.
+        //
+        // We also determine IF we need to migrate the remap shard for each collection, by
+        // concurrently reading from persist.
+        let to_create: Vec<_> = futures::stream::iter(to_migrate)
+            .map(|(id, description)| async move {
+                let dc = this
+                    .determine_remap_shard_migration(id, &description.data_source)
+                    .await;
+
+                // TODO(guswynn): determine if this is necessary if we perform all migrations before we initialize the
```

Comment:
I _think_ we can just do all of these together, after migrations. Also: it seems smart to batch all of the `register_shard_mapping` into one call, even though it's not super critical because each call is only a push to a channel. So... idk 🤷

Reply:
True...I am going to leave this as a TODO to make the semantics change of this pr as simple as possible